flink/src/main/java/org/apache/iceberg/flink/sink/BaseDeltaTaskWriter.java

```diff
@@ -69,6 +72,9 @@ public void write(RowData row) throws IOException {
 
     switch (row.getRowKind()) {
       case INSERT:
+        if (upsert) {
+          writer.delete(row);
+        }
```

Comment:
I think we could only delete row on INSERT. I don't think there will be only have `UPDATE_AFTER` row and lost `UPDATE_BEFORE` situation. @openinx please check this.

Reply:
For an update operation in flink,  the `UPDATE_AFTER` event will must be emitted to the downstream, while the `UPDATE_BEFORE` is a best effort behavior or an configured behavior from the upstream flink source.  You can take a look at this [GroupAggFunction](https://github.com/apache/flink/blob/ab70dcfa19827febd2c3cdc5cb81e942caa5b2f0/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java#L186),  if the flink source is configured to produce `UPDATE_AFTER` only, then it won't emit any `UPDATE_BEFORE` to the downstream.  

For the downstream iceberg sink,  we need to handle all the `UPDATE_AFTER`  as `UPSERT`.  That also means we need to do nothing for the `UPDATE_BEFORE` because we will remove the previous key in the next `UPDATE_AFTER` events.

Reply:
I think we can only delete row on `UPDATE_AFTER` and keep `UPDATE_BEFORE` do nothing to prevent delete one row twice

Reply:
By name, I thought INSERT is "add a new row". Then we don't need to add a delete for it. But I guess it actually means "append a row (new or updated)".

Reply:
As @openinx mentioned in https://github.com/apache/iceberg/pull/1996#issue-546072705, we need to transform INSERT/UPDATE_AFTER to be UPSERT(delete + insert). If we don't add a delete on INSERT row when upsert mode is enable, we will get duplicate rows for same primary key.